### PR TITLE
Fixes #9480 feat(nimbus): Fetch errors from FML feature inspectors for review serializer

### LINKED ISSUE #9478: Use FML in branch feature editor
Within the `RichFeatureEditor` (PageEditBranches → FormBranches → FormFeatureValue), get the app_name, versions and feature_id. If not, can we fall back to the experiment slug and feature id? Connect up the Loader to the FML endpoint and remove all temp linting.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-3871)

### LINKED ISSUE #9480: Fetch FML errors on save
Hook up the FML feature validation to the ready for review serializer so that we show FML errors in the serializer (v5) and cannot launch when FML errors exist. 

This only hooks up the save workflow. The async fetching for CodeMirror is separate ([https://mozilla-hub.atlassian.net/browse/EXP-3971](https://mozilla-hub.atlassian.net/browse/EXP-3971|smart-link))

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-3876)

### PR DESCRIPTION
**Pull request by eliserichards**
Because

- We want to hook up the FML schema validation to our save button error handling

This commit

- Fetches [FML inspectors](https://github.com/mozilla/application-services/pull/5827/files#diff-c5049ce7c2823778792aab8ac554aec8e6f63f74cd206bc70c8eb0a30aae2101R77-R85) in the NimbusFmlLoader and create a method for fetching the errors from those inspectors
- Creates a NimbusFmlLoader and fetches a list of FML errors in the ready for review serializer

Screenshot 2023-10-27 at 11 28 15 AM

https://github.com/mozilla/experimenter/assets/43795363/3845f877-4dc8-407f-ac98-4864a4bb1bcd

### CODE REVIEW

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py:255]: If you need to mock this out everywhere you can do it at the class level.  You can define a [mixin class](https://github.com/mozilla/experimenter/blob/main/experimenter/experimenter/kinto/tests/mixins.py) that sets up all the mocking and then just [inherit from it in the test class](https://github.com/mozilla/experimenter/blob/main/experimenter/experimenter/kinto/tests/test_client.py#L14) and it'll do all the mocking for all the tests.

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py:1520]: I would expect to actually validate that the FML error data we setup above is included in the response here.

**Review by jaredlockhart — CHANGES_REQUESTED** [review]
Okay cool this generally all looks correct to me!  Couple ideas to improve the tests but this is coming along great 🎉 🎉 🎉

**Review by jhugman — COMMENTED** [review]
A drive by, with a couple of questions: 

- do we need to change the type script, or is this done elsewhere?
- can we change use of schema to `text` or `json_string`?

File: experimenter/experimenter/experiments/api/v5/serializers.py
Comment (eliserichards) [reply experimenter/experimenter/experiments/api/v5/serializers.py:1450]: I can rename it something better 👍

**Comment by eliserichards** [comment]
> do we need to change the type script, or is this done elsewhere?

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py
Comment (eliserichards) [reply experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py:255]: this is exactly what I was looking for! ty!!

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py:1480]: 🙆‍♀️ 🙆‍♀️ 🙆‍♀️ 🙆‍♀️

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/mixins.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/api/v5/test_serializers/mixins.py:18]: 👏 👏 👏 👏

**Review by jaredlockhart — CHANGES_REQUESTED** [review]
Looking great and coming along nicely!  Looks like a miss somewhere:

image

**Comment by eliserichards** [comment]
So the integration tests were all failing, and when you ran them locally you got the following error in the terminal: 

```
File "/experimenter/experimenter/experiments/api/v5/__init__.py", line 3, in <module>
experimenter-experimenter-1       |     from experimenter.experiments.api.v5.mutations import Mutation
experimenter-experimenter-1       |   File "/experimenter/experimenter/experiments/api/v5/mutations.py", line 16, in <module>
experimenter-experimenter-1       |     from experimenter.experiments.api.v5.inputs import ExperimentCloneInput, ExperimentInput
experimenter-experimenter-1       |   File "/experimenter/experimenter/experiments/api/v5/inputs.py", line 4, in <module>
experimenter-experimenter-1       |     from experimenter.experiments.api.v5.types import (
experimenter-experimenter-1       |   File "/experimenter/experimenter/experiments/api/v5/types.py", line 10, in <module>
experimenter-experimenter-1       |     from experimenter.experiments.api.v5.serializers import (
experimenter-experimenter-1       |   File "/experimenter/experimenter/experiments/api/v5/serializers.py", line 35, in <module>
experimenter-experimenter-1       |     from experimenter.features.manifests.nimbus_fml_loader import NimbusFmlLoader
experimenter-experimenter-1       |   File "/experimenter/experimenter/features/manifests/nimbus_fml_loader.py", line 6, in <module>
experimenter-experimenter-1       |     from rust_fml import FmlClient, FmlFeatureInspector
experimenter-experimenter-1       | ModuleNotFoundError: No module named 'rust_fml'
```

I made the following changes and I _think_ that we are able to fetch the rust fml from the megazord correctly now: 
* `tox.ini` (for integration tests) was referencing the wrong place for AS: `application-services/components/nimbus/src` instead of `application-services/megazord`
* in the top level dockerfile, we needed to copy the AS megazord `--from=dev`
* [according to this](https://docs.docker.com/compose/environment-variables/envvars-precedence/), we weren't correctly setting the precedence for the env variable `PYTHONPATH` -- added `PYTHONPATH` to the `.env` files and the rust image dockerfile for integration tests

File: experimenter/experimenter/experiments/api/v5/serializers.py
Comment (brennie) [inline experimenter/experimenter/experiments/api/v5/serializers.py:1513]: You don't want to be interacting with the schema at all. The schema is not the FML, its a generated JSON schema.

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:103]: This should reference the GH issue.

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:55]: This should reference the GH issue.

File: experimenter/experimenter/features/tests/test_nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/tests/test_nimbus_fml_loader.py:160]: This should use a Path.

File: experimenter/experimenter/features/tests/test_nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/tests/test_nimbus_fml_loader.py:171]: here too

File: experimenter/experimenter/features/tests/test_nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/tests/test_nimbus_fml_loader.py:183]: here too

File: experimenter/experimenter/nimbus-ui/src/components/PageEditBranches/FormBranches/FormFeatureValue/RichFeatureValueEditor.tsx
Comment (brennie) [inline experimenter/experimenter/nimbus-ui/src/components/PageEditBranches/FormBranches/FormFeatureValue/RichFeatureValueEditor.tsx:94]: Shouldn't be commented out.

**Review by brennie — CHANGES_REQUESTED** [review]

**Comment by eliserichards** [comment]
@brennie this is dependent on #9639 and is not ready to be fully reviewed yet until that is merged, hence the commented out code. Sorry, should have added a comment in here.

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:103]: This is an epic with no issues (yet), so there is no GH issue

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:55]: same here

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:92]: This is broken out for testing, so we can mock the responses from the rust-fml

**Comment by eliserichards** [comment]
Ok, now that the Docker issues are all resolved, this should be ready for review 😰

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:54]: This uses the remote file path in the local repository, which is incorrect. This method is not needed.

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:77]: We talked in slack and you said you weren't using the refs because we aren't communicating with GitHub.

These should be removed because these fields from apps.yaml are being removed in an upcoming PR.

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:120]: Same note here re docstrings.

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:125]:
```
Returns:
    A list of feature manifest errors.
```

**Review by brennie — CHANGES_REQUESTED** [review]

File: experimenter/experimenter/experiments/api/v5/serializers.py
Comment (eliserichards) [reply experimenter/experimenter/experiments/api/v5/serializers.py:371]: it's already imported at the top, Pylance complains when you don't say the whole `typing.List`. It's like that in like 40 other places in this file, so I'm gonna keep this as-is

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:77]: I was going to explicitly remove them in a different PR (maybe it makes sense to remove them in the same PR that the apps.yaml is removed?)

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/mixins.py
Comment (eliserichards) [reply experimenter/experimenter/experiments/tests/api/v5/test_serializers/mixins.py:18]: I was following the precedent set by the other `mixins.py` (e.g. [`setup_kinto_pending_review`](https://github.com/mozilla/experimenter/blob/13fc2232f63c6b266ae7771c84a408ae4f6933b7/experimenter/experimenter/kinto/tests/mixins.py#L18) and [`setup_cached_sizing_data`](https://github.com/mozilla/experimenter/blob/13fc2232f63c6b266ae7771c84a408ae4f6933b7/experimenter/experimenter/jetstream/tests/mixins.py#L22C9-L22C33)). I don't think it's necessary to use `setup` instead of `set`, but it does make it internally consistent 😂 

I'd like to keep the `no_errors` separate, because we're using that in the `setUp` methods for `TestNimbusReviewSerializerSingleFeature` and `TestNimbusReviewSerializerMultiFeature` and I think it makes it a little more explicit to see `no_errors` vs `[]`. What do you think?

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:54]: same as below - will be removed separately from this PR

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:71]: #9654

File: experimenter/experimenter/experiments/api/v5/serializers.py
Comment (brennie) [reply experimenter/experimenter/experiments/api/v5/serializers.py]: `list` not `List`. As of 3.8 you can subscript list dict etc for type annotations instead of the ones exported by typing.

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:71]: Yeah I can remove these in my pr.

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/mixins.py
Comment (brennie) [reply experimenter/experimenter/experiments/tests/api/v5/test_serializers/mixins.py:18]: I was unaware of the existing mixins, so feel free to ignore this feedback :)

File: experimenter/experimenter/experiments/api/v5/serializers.py
Comment (brennie) [reply experimenter/experimenter/experiments/api/v5/serializers.py]: Also what I meant with the import was:

```python
from typing import Optional

# ...

errors: list[Optional[T]]

```

Comment (eliserichards) [reply experimenter/experimenter/experiments/api/v5/serializers.py]: Yeah, that makes sense. I'd like to keep this as-is since it's consistent with the rest of the file, though, and I'm not going to change all of the instances of this in this PR. I can file a cleanup ticket to change this 👍

**Review by jhugman — APPROVED** [review]
This is looking very nice now. @brennie do you have any final comments before Elise lands this?

Good job @eliserichards !

File: experimenter/experimenter/features/tests/test_nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/tests/test_nimbus_fml_loader.py:267]: When you do multiple patch() calls, they become arguments in reverse order, e.g.

```
@patch("foo")
@patch("bar")
@patch("baz")
def test(self, mock_baz, mock_bar, mock_foo):
```

Consider one patch call:

```
@patch("bar")
def test(self, mock_bar)
```

this is equivalent to:

```
def test(self, ...):
   ...
test = patch("bar")(test)
```

When you add a second decorator:

```
@patch("foo")
@patch("bar")
def test(self, mock_bar, mock_foo)
```

this is equivalent to 

```
def test(self, ...):
    ...
test = patch("bar")(test) # replaces test(self) with test(self, mock_bar)
test = patch("foo")(test) # replaces test(self, mock_bar) with test(self, mock_bar, mock_foo)
```

See [these docs](https://docs.python.org/3/library/unittest.mock.html#quick-guide):

> Note
>
> When you nest patch decorators the mocks are passed in to the decorated function in the same order they applied (the normal Python order that decorators are applied). This means from the bottom up, so in the example above the mock for module.ClassName1 is passed in first.
>
> With [patch()](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) it matters that you patch objects in the namespace where they are looked up. This is normally straightforward, but for a quick guide read [where to patch](https://docs.python.org/3/library/unittest.mock.html#where-to-patch).

File: experimenter/experimenter/features/tests/test_nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/tests/test_nimbus_fml_loader.py:307]: Same here, the mock arguments are swapped.

**Review by brennie — DISMISSED** [review]

File: experimenter/experimenter/experiments/api/v5/serializers.py
Comment (eliserichards) [reply experimenter/experimenter/experiments/api/v5/serializers.py:1451]: ooo much more readable. love it 🎉

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py]: oh shoot! I had these tests in my other branch, I added them in here 👍

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:93]: This should be logged at error

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:37]: The second Arg doesn't need to be an fstring. You can just reference the variable there.

**Review by brennie — DISMISSED** [review]

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:106]: Looking at the UDL, it looks like it can return None https://github.com/mozilla/application-services/blob/d58715194a98867372cdce8f75bd32b2e68b953b/components/support/nimbus-fml/src/fml.udl#L72

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (brennie) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:130]: According to the UDL this is `sequence<FmlEditorError>? get_errors(string src);` [src](https://github.com/mozilla/application-services/blob/d58715194a98867372cdce8f75bd32b2e68b953b/components/support/nimbus-fml/src/fml.udl#L72)

So `error` can never be falsy, but `inspector.get_errors()` can return None, so this should be:

```
if e := inspector.get_errors(blob):
  errors.extend(e)
```

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:10]: Yeah we're using it everywhere 👍

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (eliserichards) [reply experimenter/experimenter/features/manifests/nimbus_fml_loader.py:133]: yep, that's exactly what it's doing, except we're using `_get_errors` so we can add caching

**Review by brennie — APPROVED** [review]
r+

Sorry for all the review cycles

**Comment by eliserichards** [comment]
> r+
> 
> Sorry for all the review cycles

Don't apologize! It's super helpful 😄

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/mixins.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/api/v5/test_serializers/mixins.py:18]: 🙏 Perfect

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py:1368]: Yep these are great, good layout of cases, good assertions, all look right to me 👍

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_ready_for_review_serializer.py:2850]: Heck yeah 💯

File: experimenter/experimenter/features/manifests/nimbus_fml_loader.py
Comment (jaredlockhart) [inline experimenter/experimenter/features/manifests/nimbus_fml_loader.py:49]: Oh yep, I see what you did here, makes perfect sense 👍

**Review by jaredlockhart — APPROVED** [review]
Okay cool everything here makes sense this looks good great job @eliserichards let's validate some FML errors! 🎉 🎉 🎉

**Comment by eliserichards** [comment]
This is still blocked on `main` being broken

**Comment by eliserichards** [comment]
Waiting on https://github.com/mozilla/experimenter/pull/9712

**Pull request merged**
